File: android/src/main/java/com/textrecognition/TextRecognitionTypes.kt
package com.textrecognition

import com.facebook.react.bridge.ReadableMap

public class RNTROptions {
    // mlkit doesnt seem to have any options

    public fun from(options: ReadableMap) {
        // mlkit doesnt seem to have any options
    }
}

public class ImageProperties {
    public var ColorModel: String = "";
    public var DPIHeight: Int = 0;
    public var DPIWidth: Int = 0;
    public var HasAlpha: Boolean = false;
    public var PixelHeight: Int = 0;
    public var PixelWidth: Int = 0;
//    public var Exif: Exif = Exif();
//    public var Jfif: Jfif = Jfif();
//    public var Tiff: Tiff = Tiff();
//    public var PNG: PNG = PNG();
}

//public class Exif {
//    public var ColorSpace: Int = 0;
//    public var PixelXDimension: Int = 0;
//    public var PixelYDimension: Int = 0;
//}
//
//public class Jfif {
//    public var DensityUnit: Int = 0;
//    public var JFIFVersion: List<Int> = listOf<Int>();
//    public var XDensity: Int = 0;
//    public var YDensity: Int = 0;
//}
//
//public class Tiff {
//    public var Orientation: Int = 0;
//}
//
//public class PNG {
//    public var InterlaceType: Int = 0;
//    public var XPixelsPerMeter: Int = 0;
//    public var YPixelsPerMeter: Int = 0;
//}
public class StringProperties {
    public var text: String = "";
    public var languageCode: String = "";
    public var confidence = 0.0F;
    public var leftX = 0;
    public var middleX = 0;
    public var rightX = 0;
    public var bottomY = 0;
    public var middleY = 0;
    public var topY = 0;
    public var width = 0;
    public var height = 0;
}

public class ReturnValue {
    public var imageProperties: ImageProperties = ImageProperties();
    public var stringProperties: MutableList<StringProperties> = mutableListOf<StringProperties>();
}

File: ios/TextRecognitionTypes.swift
//
//  TextRecognitionTypes.swift
//  TextRecognition
//
//  Created by Joey Eamigh on 4/24/23.
//  Copyright © 2023 Facebook. All rights reserved.
//

import Foundation
import NaturalLanguage
import Vision

extension String {
  func stripPrefix(_ prefix: String) -> String {
    guard hasPrefix(prefix) else { return self }
    return String(dropFirst(prefix.count))
  }
}

enum RecognitionLevel: String, Codable {
  case fast = "fast"
  case accurate = "accurate"
}

struct RNTROptions {
  var visionIgnoreThreshold = 0.0
  var automaticallyDetectLanguage = true
  var recognitionLanguages: [String] = ["en-US"]
  var useLanguageCorrection = false
  var customWords: [String] = []
  var recognitionLevel: RecognitionLevel = .accurate
}

extension RNTROptions: Codable {
  init(dictionary: NSDictionary) throws {
    self.visionIgnoreThreshold = dictionary["visionIgnoreThreshold"] as? Double ?? 0.0
    self.automaticallyDetectLanguage = dictionary["automaticallyDetectLanguage"] as? Bool ?? true
    self.recognitionLanguages = dictionary["recognitionLanguages"] as? [String] ?? ["en-US"]
    self.useLanguageCorrection = dictionary["useLanguageCorrection"] as? Bool ?? false
    self.customWords = dictionary["customWords"] as? [String] ?? []
    self.recognitionLevel =
      RecognitionLevel(rawValue: dictionary["recognitionLevel"] as? String ?? "accurate")
      ?? .accurate
  }
}

struct StringProperties {

}

struct Exif {
  let ColorSpace: Int
  let PixelXDimension: Int
  let PixelYDimension: Int
}

struct Jfif {
  let DensityUnit: Int
  let JFIFVersion: [Int]
  let XDensity: Int
  let YDensity: Int
}

struct Tiff {
  let Orientation: Int
}

struct PNG {
  let InterlaceType: Int
  let XPixelsPerMeter: Int
  let YPixelsPerMeter: Int
}

struct ImageProperties {
  let ColorModel: String
  let Depth: Int
  let DPIHeight: Int
  let DPIWidth: Int
  let HasAlpha: Bool
  let Orientation: Int
  let PixelHeight: Int
  let PixelWidth: Int
  let ProfileName: String
  let Exif: Exif
  let JFIF: Jfif
  let TIFF: Tiff
  let PNG: PNG
}

//struct RecognitionResolve {
//  let stringProperties
//  let imageProperties
//}
